### DISCUSSION TITLE
Crystal Plasticity with Periodic Boundary Conditions

### DISCUSSION THREAD

**Discussion by lkuna24** [idaholab/moose#28778] [Q&A Modules: Solid mechanics]
I am attempting to run crystal plasticity simulations quasi-periodic boundary conditions.  I am applying a displacement to the top/bottom (z) of my RVE while setting the in-plane (x + y) boundaries to be periodic.  For some reason, the PBC simulation shows a very slight yield but otherwise almost linear curve.  To test this, I ran the same simulation utilizing symmetric boundary conditions (u_i = 0 on one of the boundaries normal to displacement), and now I see an expected result.  Here is a plot of the two simulations:

![BCsTest](https://github.com/user-attachments/assets/2b2de954-0418-4a38-b0d7-0b27ec2e589a)

I have included the input file, mesh, and txt to seed the orientations on the blocks.  

Additionally, I know that the GlobalStrain system has been implemented for SolidMechanics PBC simulations and have also constructed an input for that.  This has an extremely hard time converging, and even looking at the result is quite strange.  It seems to have completed flattened the RVE.  Maybe using GS is the solution to this issue, and maybe I have just not utilized it properly?

![GS_Flattening](https://github.com/user-attachments/assets/8f5746f3-6c81-4f85-ad84-5b78df14a367)

Any advice is greatly appreciated.

Thanks for your time and attention.

[DRAGen_RVE_16G.zip](https://github.com/user-attachments/files/17263312/DRAGen_RVE_16G.zip)
[Periodic_CP_Test.txt](https://github.com/user-attachments/files/17263320/Periodic_CP_Test.txt)
[EulerAngles.txt](https://github.com/user-attachments/files/17263324/EulerAngles.txt)
[Periodic_CP_GS_Test.txt](https://github.com/user-attachments/files/17263334/Periodic_CP_GS_Test.txt)

**Comment by sapitts** [comment]
I'm not aware of the periodic boundary conditions having been successfully used with the crystal plasticity models in MOOSE: I am not certain that the assumptions in the CP models are consistent with the PBC assumptions. @SudiptaBiswas, @hugary1995, or @dschwen, do you know?

Note: if you would like to ask for one of the moose developers to review your input files and/or code, a github branch is the best option. Many of us are unable to download zip files from the internet

**Reply by lkuna24** [reply]
Apologies, I should have known.  
I have pushed all the files to a lone branch located here:
https://github.com/lkuna24/moose_CP_files.git

**Reply by hugary1995** [reply]
I suggest you set
ux=0 on left (-x)
uy=0 on bottom (-y)
uz=0 on back (-z)
uz=g on front (+z)

Also apply an equal value boundary constraint on front (+z).

**Reply by lkuna24** [reply]
Thanks Gary, but I noted in my original post that this is how I first checked as to whether or not the PBC with CP is working:

> boundary conditions (u_i = 0 on one of the boundaries normal to displacement)

I should have been a little bit clearer about this.

**Reply by hugary1995** [reply]
Okay, good.

Then for the periodic case, do you want to apply displacement periodicity or strain periodicity?

If you do displacement periodicity, I believe the plastic spin is almost zero, hence little hardening.

**Reply by lkuna24** [reply]
I would hope to conserve the strain periodicity.  As it is an AuxVariable (I believe), I think the PBC has to be specified on displacements (the Variables in this simulation).

**Reply by hugary1995** [reply]
The two are not compatible in general.

**Reply by lkuna24** [reply]
Right, so is there a solution to resolve this?  I was under the impression that Global Strain was implemented to conserve strain periodicity.  Maybe the way I have utilized the GlobalStrain Action is just incorrect here?

**Comment by lkuna24** [comment]
Just checking in on this post again.  Is it currently not possible to do a CP RVE simulation with PBC in MOOSE?
